Retry sending notifications (pulses) on failure

metamben opened the pull request:
Fixes #12487.

There are still some issues:

- [x] There are two parts, one for emails and another doing more or less the same for Slack messaging. They are currently duplicates. I can try extracting the common things somehow, but I'm not sure the extraction would be worth the effort.
- [x] The configuration is fairly complicated: the settings have an update function that gets overridden (the linter doesn't like this). This is because there is circular dependence between the update function and the settings. The whole point of the update function is to be able to define the retry object whenever the settings change. We can get rid of this by defining a new retry object on every invocation. (The retry object accumulates information about retries, which might be useful.)
- [x] I'm looking for an elegant way to define default settings for tests/dev. Some tests trigger emails and I would prefer not retrying for long to be the default.
  - [x] Find a way to prevent existing DB settings overwrite the defaults during testing. (I had a DB with 5 retries and every time I tested in the REPL the values in the DB overrode the defaults which made some tests (e.g., the ones in `metabase.api.user-test`) fail.)

metamben committed:
Retry sending emails on failure

dpsutton commented on src/metabase/email.clj:
```diff
@@ -129,6 +130,70 @@
       (log/warn e (trs "Failed to send email"))
       {::error e})))
 
+(defn- reconfigure-retrying "No-op. Will be overridden." [_ _])
```
should this just be `(declare reconfigure-retrying)`? Or is the private metadata messing with it?

metamben replied:
The problem is, this function is getting called before the real definition is read. It seems when the settings get their initial values, the update function gets called. (Which I think is good thing.)

metamben replied:
Unfortunately, the current "solution" is not just ugly, it triggers the linter too.

noahmoss replied:
Are you sure? I tried replacing this with `declare` locally and everything seems to work as expected.

noahmoss commented:
Some initial thoughts...
> It can duplicate the configuration part or I can try extracting the common things somehow

I think separate configs for email vs Slack makes sense. I also can see an argument for making the retry config _simpler_ — for example, only allowing `max-attempts` and `multiplier` knobs to be configured, and having static values for the rest. I can't really think of a reason someone would need to tweak the other values.

Then again, it probably doesn't hurt to expose them all as env variables. If we were adding these to the UI I'd definitely limit it to just those two values. (But we'd also want to get input from product before doing that.)

> We can get rid of this by defining a new retry object on every invocation

I think this would be fine. Some basic timing of `(make-retry-state)` on my machine is showing it consistently run under 1 ms, which is marginal compared to the time it takes to actually send an email or Slack message. And sending messages is (usually) an async operation anyway so we don't need to worry too much about responsiveness.

> I'm looking for an elegant way to define default settings for tests/dev

I think it would make sense to disable retries entirely for most email tests (aside from ones that are explicitly testing the retry functionality). Could you modify the email test helper (`metabase.email-test/with-fake-inbox`) to set `email-retry-max-attempts` to 1 by default?

metamben commented:
>  it consistently run under 1 ms, which is marginal compared to the time it takes to actually send an email or Slack message

Right, I agree that this should not become a bottleneck. On the other hand, if we do this, we lose the state (which is mostly statistics). Since the state/statistics are not exposed this is not a big loss. It bothers me more that using fresh retry objects on each invocation is not the intended way of using the library and I wouldn't even consider this if we didn't have the cyclic dependency in the configuration. In other words, I would very much prefer to solve the configuration problem somehow and use the library as intended.

> Could you modify the email test helper (metabase.email-test/with-fake-inbox) to set email-retry-max-attempts to 1 by default?

This is a good idea, will look into it. Unfortunately, the Slack integration uses HTTP and the tests rely on `http-fake/with-fake-routes`, so I'm not sure the same idea can be applied there. I wonder if `defsetting` can handle these defaults or if there is already a way to set default environment variables, or something similar.

noahmoss commented:
> On the other hand, if we do this, we lose the state (which is mostly statistics).

I see, that's an interesting point (I wasn't already familiar with the resilience4j library so not aware of the full capabilities). 

Here's another option: use a single JSON setting for the retry config, rather than multiple individual settings. Then, `reconfigure-retrying` can use its `new-value` argument to build a new retry object, and doesn't have to re-read the settings. The downside to this is that configuring the values via env variable is less straightforward but we do have precedent for stuff like this already (search [this page](https://www.metabase.com/docs/latest/operations-guide/environment-variables.html#mb_jwt_group_mappings) for "json" to see examples).

> I wonder if defsetting can handle these defaults

You should be able to use `metabase.test.util/with-temporary-setting-values` to override any settings in tests, for the duration of the test. In the Slack case you'd just have to do this in each test, or write a new test helper macro to abstract it, similar to the email tests.

metamben commented:
> use a single JSON setting for the retry config

I have considered this, but as you pointed out yourself, most people will probably want to change a couple of values, not all of them.

> You should be able to use metabase.test.util/with-temporary-setting-values to override any settings in tests, for the duration of the test.

I'm doing this right now. The problem is, even tests that don't care about emails sometimes send emails in the background which can get pretty ugly.

> Also, FYI — we have is-dev?, is-prod? and is-test? bindings in metabase.config 

Thanks for the hint, I'll try something with these.

noahmoss commented:
> most people will probably want to change a couple of values, not all of them

Well, we could provide defaults for any values that a user doesn't set in the JSON; they wouldn't have to set all the values at once. My gut tells me that would be fine, but it's your call.

(also cc @flamber -- any thoughts on whether a JSON config for email retry settings would be too onerous for customers?)

metamben committed:
Retry posting slack messages on failure

escherize commented on src/metabase/email.clj:
```diff
@@ -129,6 +131,74 @@
+                :randomization-factor (email-retry-randomizaion-factor)
+                :max-interval-millis (email-retry-max-interval-millis)
+                :retry-on-result-pred #(and (map? %) (instance? Throwable (::error %)))})]
+    {:retry retry
```
I don't see retry getting used anywhere (besides in sender below), so is it needed in this map? 

Hmm, it could be helpful for debugging though. Is that right?

metamben replied:
> I don't see retry getting used anywhere (besides in sender below), so is it needed in this map?

Right, it's not currently used, I just thought that if we don't create one on every call, then why not keep a reachable reference to it.  (There is another [discussion](https://github.com/metabase/metabase/pull/21881#issuecomment-1111195584) about the retry object.)

noahmoss commented on test/metabase/test/util.clj:
```diff
@@ -1162,3 +1162,12 @@
       (.read is ary)
       ary)))
+
+(defn works-after
+  [x f]
```
You'll need a docstring on here to appease the linter

noahmoss reviewed (approved):
LGTM overall! I'm still not totally sure about the `reconfigure-retrying` stuff...it seems to me that using `declare` up-front works locally but maybe I'm missing something.

We also have a very naive `auto-retry` macro in `metabase.util`...would it be possible to change that to also use this new retry lib? It could create a basic retry object on the fly and use it... or if that's too heavyweight, maybe just add a comment to the `auto-retry` macro pointing out that these retry utils also exist, and what the differences are?

metamben commented:
> I tried replacing this with declare locally and everything seems to work as expected.

I've retested and indeed cannot reproduce the error I got earlier. (Maybe that's because of rebasing onto the master or maybe I had a REPL in some weird state.) Thanks for spotting this!

metamben committed:
Use declare for var getting redefined later

metamben committed:
Beautify works-after

metamben committed:
Make the Slack client retry a bit more aggressively by default

metamben committed:
Support categorizing exceptions as retying/fatal

Also extend the documentation of metabase.util/auto-retry to link to the new
namespace.

flamber reviewed (commented):
Could this be simplified a bit, so there's not specific settings for email and slack, but just a single set for "notifications"?

Since this is just a protection against connection issues - not a full blown stored queue - then I think it's fine how the default email settings are made (if I understand them correctly). And probably not even need settings - perhaps just a `notification-retry-max-attempts`, which let you disable it with `0`.

I'm guessing that Metabase will try to send an email, on failure it will retry on the seconds 15, 30, 60, 120 and finally 240 after which it gives up (and provides a log)?
That seems very https://github.com/metabase/metabase/blob/master/zen.md

I would be very glad for settings if we had a real queue, but I think this is fine to be kept very simple so it works for most - and people who needs more would use Postfix etc to queue emails for days if needed.

metamben commented:
> Could this be simplified a bit, so there's not specific settings for email and slack, but just a single set for "notifications"?

We could do that, but I wouldn't like that much. For example, just because I have a good solution for emails and don't want to make (many) retries, why should I have to do the same for slack? Also, I have defined different defaults for slack, because I think people's expectation about the responsiveness/delays of slack and email are different. (My expectations are definitely different.)

Now, we could still choose the middle ground and make just the number of retries configurable for both email and slack, but in fact it's quite difficult to find defaults that fit 90% of the users. (I'm not sure at all we have found them.) Is it a big cognitive load to have five settings instead of one even if you don't _have to_ use all, do you think?

metamben committed:
Force no retry form emails and slack in dev and test modes

metamben committed:
Use reconfigure-retrying directly instead of through the var

This is OK because the var is now bound only once.

metamben committed:
Prevent email and slack retries during Cypress tests

metamben committed:
Restrict retrying to pulse notifications

This is to reduce the risk of breaking the UI by blocking threads and
running into timeouts.

noahmoss reviewed (approved):
Nice, the revised approach of just wrapping `send-notification!` in retries looks good.

metamben committed:
Fix typo in setting documentation

flamber commented on src/metabase/pulse.clj:
```diff
@@ -373,22 +377,96 @@
+  :default 2.0
+  :on-change reconfigure-retrying)
+
+(defsetting notification-retry-randomizaion-factor
```
@metamben 
We missed this misspelling. If we fix it, can we drop the old setting, so it won't cause problems during Serialization (unknown settings causes errors on load)?

randomizaion => randomization

dpsutton replied:
oh good catch. We can fix the spelling of the setting and then
- if the correctly spelled version is null, look for the misspelled version
- if present and its not an env variable value, clear it and set the correctly spelled version
- if its an env variable we cannot clear its value. We can warn that we've fixed the misspelling and they should fix the mispelling in their startup command for metabase

flamber replied:
If we fix it soon-ish, then I think it's okay to completely ignore the old spelling in envvar, since it's currently only active on 44.0, so I don't think a lot has used this envvar yet. That should simplify the fix too.
